File: template_math.py
# BIT MANIPULATION TRICKS
# a | b = a ^ b + a & b
# a ^ (a & b) = (a | b) ^ b
# b ^ (a & b) = (a | b) ^ a
# (a & b) ^ (a | b) = a ^ b
# a + b = a | b + a & b
# a + b = a ^ b + 2 * (a & b)
# a - b = (a ^ (a & b)) - ((a | b) ^ a)
# a - b = ((a | b) ^ b) - ((a | b) ^ a)
# a - b = (a ^ (a & b)) - (b ^ (a & b))
# a - b = ((a | b) ^ b) - (b ^ (a & b))
# if x is a power of two ---> x&(x-1) == 0

### GAUSS ELIMINATION

def gauss_solve(A, b):
    """
    Solve A x = b using Gaussian elimination.
    
    Input:
        A : list of lists (n × m matrix)
        b : list (n vector)

    Returns:
        (status, solution)
            status = 1  -> unique solution
            status = 2  -> infinite solutions
            status = -1 -> no solution
            solution = list of variable values if exist
    
    Author: akshitsinghvi28
    """

    a = [row + [b[i]] for i, row in enumerate(A)]

    eps = 1e-9
    n = len(a)
    m = len(a[0]) - 1
    pos = [-1] * m
    rank = 0

    for col in range(m):
        row = rank
        mx = row

        for i in range(row, n):
            if abs(a[i][col]) > abs(a[mx][col]):
                mx = i

        if abs(a[mx][col]) < eps:
            continue

        a[row], a[mx] = a[mx], a[row]
        pos[col] = row

        for i in range(n):
            if i != row and abs(a[i][col]) > eps:
                c = a[i][col] / a[row][col]
                for j in range(col, m+1):
                    a[i][j] -= a[row][j] * c

        rank += 1

    ans = [0] * m
    for i in range(m):
        if pos[i] != -1:
            ans[i] = a[pos[i]][m] / a[pos[i]][i]

    for i in range(n):
        s = sum(ans[j] * a[i][j] for j in range(m))
        if abs(s - a[i][m]) > eps:
            return -1, []

    for i in range(m):
        if pos[i] == -1:
            return 2, ans

    return 1, ans


### MATRIX EXPONENTIATION
from typing import List, Optional, Union

MOD = 10**9 + 7
L = 3  # example size; assumed to be defined globally

class Mat:
    def __init__(
        self,
        init: Optional[Union["Mat", List[List[int]]]] = None
    ) -> None:
        # initialize zero matrix
        self.a: List[List[int]] = [[0] * L for _ in range(L)]

        if init is None:
            return

        if isinstance(init, Mat):
            # copy from another Mat
            for i in range(L):
                for j in range(L):
                    self.a[i][j] = init.a[i][j]

        elif isinstance(init, list):
            # initialize from list of lists
            if len(init) != L or any(len(row) != L for row in init):
                raise ValueError(f"Expected a {L}x{L} matrix")

            for i in range(L):
                for j in range(L):
                    self.a[i][j] = init[i][j] % MOD

        else:
            raise TypeError("Mat can be initialized with Mat, list of lists, or None")

    def __mul__(self, other: "Mat") -> "Mat":
        result = Mat()
        for i in range(L):
            for j in range(L):
                for k in range(L):
                    result.a[i][j] = (
                        result.a[i][j] + self.a[i][k] * other.a[k][j]
                    ) % MOD
        return result

    def transpose(self) -> "Mat":
        """Return the transpose of the matrix as a new Mat object."""
        result = Mat()
        for i in range(L):
            for j in range(L):
                result.a[j][i] = self.a[i][j]
        return result

def I() -> Mat:
    m = Mat()
    for i in range(L):
        m.a[i][i] = 1
    return m

def quickmul(x: Mat, y: int) -> Mat:
    ans = I()
    cur = x
    while y:
        if y & 1:
            ans = ans * cur
        cur = cur * cur
        y >>= 1
    return ans



class Combinatorics:
    """
    Efficient nCr (binomial coefficient) modulo MOD with precomputation.

    Precomputes factorials up to N, so each nCr query is O(1).
    Assumes MOD is prime (default = 1e9+7).
    """

    def __init__(self, N: int, 
            MOD: int = 10**9 + 7
            # MOD: int = 998244353    
        ):
        self.N = N
        self.MOD = MOD
        self.fact = [1] * (N + 1)
        for i in range(1, N + 1):
            self.fact[i] = (self.fact[i - 1] * i) % self.MOD
        self.d = self._get_derangements(N)
    
    def _inversemod(self, a: int) -> int:
        """Return modular inverse of a modulo MOD (MOD must be prime)."""
        return pow(a, self.MOD - 2, self.MOD)

    def nCr(self, n: int, r: int) -> int:
        """Return nCr % MOD, or 0 if r > n."""
        if r > n: return 0
        p1 = (self.fact[n] * self._inversemod(self.fact[r])) % self.MOD
        p2 = self._inversemod(self.fact[n - r])
        return (p1 * p2) % self.MOD

    # Function to calculate derangements !n
    # !n=(n−1)(!(n−1)+!(n−2)) where !0 = 1 and !1 = 0
    # Time complexity: O(n)
    def _get_derangements(self, n):
        """Return derangements upto n."""
        if n == 0: return 1
        if n == 1: return 0
        dp = [0] * (n + 1)
        dp[0] = 1
        dp[1] = 0
        for i in range(2, n + 1): dp[i] = ((i - 1) * (dp[i - 1] + dp[i - 2])) % MOD
        return dp

    def derangement(self, n): 
        return self.d[n]

class ExactCombinatorics:
    def __init__(self, N: int):
        self.N = N
        self.fact = [1] * (N + 1)
        for i in range(1, N + 1):
            self.fact[i] = self.fact[i - 1] * i
        self.d = self._get_derangements(N)

    def nCr(self, n: int, r: int) -> int:
        """Compute exact n choose r."""
        if r > n or n < 0 or r < 0:
            return 0
        return self.fact[n] // (self.fact[r] * self.fact[n - r])

    def _get_derangements(self, n: int):
        """Precompute derangements exactly (no mod)."""
        if n == 0:
            return [1]
        if n == 1:
            return [1, 0]
        dp = [0] * (n + 1)
        dp[0] = 1
        dp[1] = 0
        for i in range(2, n + 1):
            dp[i] = (i - 1) * (dp[i - 1] + dp[i - 2])
        return dp

    def derangement(self, n: int) -> int:
        """Return exact derangement count for n."""
        if n > self.N:
            raise ValueError(f"Precomputed only up to {self.N}")
        return self.d[n]


class Basis:
    """Linear basis for XOR over integers."""
    
    def __init__(self, bit_width=31):
        """Initialize empty basis with given bit width."""
        self.B = bit_width
        self.basis = [0] * self.B
        self.sz = 0

    def clear(self):
        """Reset the basis to empty."""
        self.basis = [0] * self.B
        self.sz = 0

    def insert(self, x):
        """Insert a number into the XOR basis."""
        for i in range(self.B - 1, -1, -1):
            if (x >> i) & 1:
                if self.basis[i]:
                    x ^= self.basis[i]
                else:
                    self.basis[i] = x
                    self.sz += 1
                    break

    def can(self, x):
        """Check if some subset XOR equals x."""
        for i in range(self.B - 1, -1, -1):
            x = min(x, x ^ self.basis[i])
        return x == 0

    def max_xor(self, x=0):
        """Return maximum XOR obtainable with optional initial x."""
        for i in range(self.B - 1, -1, -1):
            x = max(x, x ^ self.basis[i])
        return x

    def kth(self, k):
        """Return the k-th smallest subset XOR (1-indexed)."""
        if k < 1 or k > (1 << self.sz):
            return -1
        
        x = 0
        cnt = (1 << self.sz)
        
        for i in range(self.B - 1, -1, -1):
            if self.basis[i]:
                limit = cnt >> 1
                if k > limit:
                    if not ((x >> i) & 1):
                        x ^= self.basis[i]
                    k -= limit
                else:
                    if (x >> i) & 1:
                        x ^= self.basis[i]
                cnt >>= 1
        return x

    def count_lt(self, x):
        """Count number of subset XOR values strictly less than x."""
        if x < 0:
            return 0
        
        ans = 0
        cnt = (1 << self.sz)
        mask = 0
        
        for i in range(self.B - 1, -1, -1):
            if self.basis[i]:
                half_cnt = cnt >> 1
                if (x >> i) & 1:
                    ans += half_cnt
                    if not ((mask >> i) & 1):
                        mask ^= self.basis[i]
                else:
                    if (mask >> i) & 1:
                        mask ^= self.basis[i]
                cnt >>= 1
            else:
                if ((x >> i) & 1) != ((mask >> i) & 1):
                    if (x >> i) & 1:
                        return ans + cnt
                    else:
                        return ans
        return ans

File: cpp-templates/template_math.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;
const ll MOD = 1000000007LL;

tuple<int, vector<double>> gauss_solve(vector<vector<double>> A, vector<double> b) {
    const double eps = 1e-9;
    int n = (int)A.size();
    int m = A.empty() ? 0 : (int)A[0].size();
    vector<vector<double>> a(n, vector<double>(m + 1));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) a[i][j] = A[i][j];
        a[i][m] = b[i];
    }

    vector<int> pos(m, -1);
    int rank = 0;
    for (int col = 0; col < m; col++) {
        int row = rank, mx = row;
        for (int i = row; i < n; i++) {
            if (fabs(a[i][col]) > fabs(a[mx][col])) mx = i;
        }
        if (mx >= n || fabs(a[mx][col]) < eps) continue;
        swap(a[row], a[mx]);
        pos[col] = row;

        for (int i = 0; i < n; i++) {
            if (i != row && fabs(a[i][col]) > eps) {
                double c = a[i][col] / a[row][col];
                for (int j = col; j <= m; j++) a[i][j] -= a[row][j] * c;
            }
        }
        rank++;
    }

    vector<double> ans(m, 0);
    for (int i = 0; i < m; i++) {
        if (pos[i] != -1) ans[i] = a[pos[i]][m] / a[pos[i]][i];
    }

    for (int i = 0; i < n; i++) {
        double s = 0;
        for (int j = 0; j < m; j++) s += ans[j] * a[i][j];
        if (fabs(s - a[i][m]) > eps) return {-1, {}};
    }
    for (int i = 0; i < m; i++) {
        if (pos[i] == -1) return {2, ans};
    }
    return {1, ans};
}

template <int L, ll MODV = MOD>
struct Mat {
    array<array<ll, L>, L> a{};

    Mat() {
        for (int i = 0; i < L; i++) {
            for (int j = 0; j < L; j++) a[i][j] = 0;
        }
    }

    static Mat identity() {
        Mat m;
        for (int i = 0; i < L; i++) m.a[i][i] = 1;
        return m;
    }

    Mat operator*(const Mat& other) const {
        Mat result;
        for (int i = 0; i < L; i++) {
            for (int k = 0; k < L; k++) {
                if (!a[i][k]) continue;
                for (int j = 0; j < L; j++) {
                    result.a[i][j] = (result.a[i][j] + a[i][k] * other.a[k][j]) % MODV;
                }
            }
        }
        return result;
    }

    Mat transpose() const {
        Mat result;
        for (int i = 0; i < L; i++) {
            for (int j = 0; j < L; j++) result.a[j][i] = a[i][j];
        }
        return result;
    }
};

template <int L, ll MODV = MOD>
Mat<L, MODV> quickmul(Mat<L, MODV> x, long long y) {
    Mat<L, MODV> ans = Mat<L, MODV>::identity();
    while (y) {
        if (y & 1) ans = ans * x;
        x = x * x;
        y >>= 1;
    }
    return ans;
}

ll mod_pow(ll a, ll e, ll mod) {
    ll res = 1 % mod;
    a %= mod;
    while (e) {
        if (e & 1) res = (__int128)res * a % mod;
        a = (__int128)a * a % mod;
        e >>= 1;
    }
    return res;
}

struct Combinatorics {
    int N;
    ll mod;
    vector<ll> fact, derangements;

    Combinatorics(int N = 0, ll mod = MOD) { init(N, mod); }

    void init(int N_, ll mod_) {
        N = N_;
        mod = mod_;
        fact.assign(N + 1, 1);
        for (int i = 1; i <= N; i++) fact[i] = fact[i - 1] * i % mod;
        derangements = get_derangements(N);
    }

    ll inversemod(ll a) const {
        return mod_pow(a, mod - 2, mod);
    }

    ll nCr(int n, int r) const {
        if (r > n || n < 0 || r < 0) return 0;
        return fact[n] * inversemod(fact[r]) % mod * inversemod(fact[n - r]) % mod;
    }

    vector<ll> get_derangements(int n) const {
        vector<ll> dp(n + 1, 0);
        if (n >= 0) dp[0] = 1;
        if (n >= 1) dp[1] = 0;
        for (int i = 2; i <= n; i++) dp[i] = (i - 1LL) * (dp[i - 1] + dp[i - 2]) % mod;
        return dp;
    }

    ll derangement(int n) const {
        return derangements[n];
    }
};

struct Basis {
    int B, sz;
    vector<long long> basis;

    Basis(int bit_width = 31) : B(bit_width), sz(0), basis(bit_width, 0) {}

    void clear() {
        fill(basis.begin(), basis.end(), 0);
        sz = 0;
    }

    void insert(long long x) {
        for (int i = B - 1; i >= 0; i--) {
            if ((x >> i) & 1LL) {
                if (basis[i]) x ^= basis[i];
                else {
                    basis[i] = x;
                    sz++;
                    break;
                }
            }
        }
    }

    bool can(long long x) const {
        for (int i = B - 1; i >= 0; i--) x = min(x, x ^ basis[i]);
        return x == 0;
    }

    long long max_xor(long long x = 0) const {
        for (int i = B - 1; i >= 0; i--) x = max(x, x ^ basis[i]);
        return x;
    }

    long long kth(long long k) const {
        if (k < 1 || k > (1LL << sz)) return -1;
        long long x = 0, cnt = 1LL << sz;
        for (int i = B - 1; i >= 0; i--) {
            if (!basis[i]) continue;
            long long limit = cnt >> 1;
            if (k > limit) {
                if (!((x >> i) & 1LL)) x ^= basis[i];
                k -= limit;
            } else {
                if ((x >> i) & 1LL) x ^= basis[i];
            }
            cnt >>= 1;
        }
        return x;
    }

    long long count_lt(long long x) const {
        if (x < 0) return 0;
        long long ans = 0, cnt = 1LL << sz, mask = 0;
        for (int i = B - 1; i >= 0; i--) {
            if (basis[i]) {
                long long half = cnt >> 1;
                if ((x >> i) & 1LL) {
                    ans += half;
                    if (!((mask >> i) & 1LL)) mask ^= basis[i];
                } else if ((mask >> i) & 1LL) {
                    mask ^= basis[i];
                }
                cnt >>= 1;
            } else if (((x >> i) & 1LL) != ((mask >> i) & 1LL)) {
                return ((x >> i) & 1LL) ? ans + cnt : ans;
            }
        }
        return ans;
    }
};
